go/service/gregor.go

```diff
@@ -1169,21 +1171,46 @@ func (g *gregorHandler) auth(ctx context.Context, cli rpc.GenericClient) (err er
 }
 
 func (g *gregorHandler) pingLoop() {
+
+	duration := g.G().Env.GetGregorPingInterval()
+	timeout := g.G().Env.GetGregorPingTimeout()
+	g.Debug("ping loop: starting up: duration: %v timeout: %v", duration, timeout)
+	defer g.Debug("ping loop: terminating")
+
 	for {
 		select {
-		case <-g.G().Clock().After(g.G().Env.GetGregorPingInterval()):
-			_, err := gregor1.IncomingClient{Cli: g.cli}.Ping(context.Background())
+		case <-g.G().Clock().After(duration):
+
+			if !g.IsConnected() {
+				g.Debug("ping loop: skipping ping since not connected")
+				continue
+			}
+
+			ctx, cancel := context.WithTimeout(context.Background(), timeout)
+			_, err := gregor1.IncomingClient{Cli: g.cli}.Ping(ctx)
+			cancel()
 			if err != nil {
-				g.Warning("error in ping loop: %s", err)
+				if err == ErrGregorTimeout {
+					g.Debug("ping loop: timeout: terminating connection")
+					g.Shutdown()
+
+					if err := g.Connect(g.uri); err != nil {
+						g.Debug("ping loop: error connecting: %s", err.Error())
+					}
```

Comment:
After thinking about it some more, I think you are right, this is too risky. I put in a change which makes it so that we never exit the ping loop.